README.md
# openapi
ICure Medical Device Dart SDK

## Requirements

Dart 2.12 or later

## Installation & Usage

### Pub.dev
The recommended way to get this package is through pub.dev, add the following dependency to your pubspec.yaml
```
dependencies:
  icure_medical_device_dart_sdk:
    path: /path/to/openapi
```

### Github
If you wish to get this package from Github, add the following dependency to your pubspec.yaml
```
dependencies:
  icure_medical_device_dart_sdk: ^1.4.4
```

## Tests

TODO

## Getting Started

Please follow the [installation procedure](#installation--usage) and then run the following:

```dart
import 'package:icure_medical_device_dart_sdk/api.dart';

final api = MedTechApiBuilder()
    .withICureBasePath('https://kraken.icure.dev')
    .withUserName('user')
    .withPassword('password')
    .addKeyPair('id', private_key)
.build();

final coding_api = api.codingApi;
final coding = Coding(id: "typ|123|1");

try {
    final result = coding_api.createOrModifyCoding(coding);
    print(result);
} catch (e) {
    print('Exception when calling CodingApi->createOrModifyCoding: $e\n');
}

```

## Documentation for API Endpoints

All URIs are relative to *http://127.0.0.1:8912*

Class | Method | Description
------------ | ------------- | -------------
*CodingApi* | [**createOrModifyCoding**](doc//CodingApi.md#createormodifycoding) | Create or update a [Coding]
*CodingApi* | [**createOrModifyCodings**](doc//CodingApi.md#createormodifycodings) | Create or update a batch of [Coding]
*CodingApi* | [**filterCoding**](doc//CodingApi.md#filtercoding) | Load codings from the database by filtering them using the provided [filter].
*CodingApi* | [**getCoding**](doc//CodingApi.md#getcoding) | Get a [Coding]
*CodingApi* | [**matchCoding**](doc//CodingApi.md#matchcoding) | Load coding ids from the database by filtering them using the provided [filter].
*DataSampleApi* | [**createOrModifyDataSampleFor**](doc//DataSampleApi.md#createormodifydatasamplefor) | Create or update a [DataSample] for a patient
*DataSampleApi* | [**createOrModifyDataSamplesFor**](doc//DataSampleApi.md#createormodifydatasamplesfor) | Create or update a batch of [DataSample] for a patient
*DataSampleApi* | [**deleteAttachment**](doc//DataSampleApi.md#deleteattachment) | Delete an attachment of a DataSample
*DataSampleApi* | [**deleteDataSample**](doc//DataSampleApi.md#deletedatasample) | Delete a [DataSample] by its id
*DataSampleApi* | [**deleteDataSamples**](doc//DataSampleApi.md#deletedatasamples) | Delete a batch of [Data Samples]
*DataSampleApi* | [**filterDataSample**](doc//DataSampleApi.md#filterdatasample) | Find data samples using the provided [filter].
*DataSampleApi* | [**getDataSample**](doc//DataSampleApi.md#getdatasample) | Get a [DataSample] by its id
*DataSampleApi* | [**getDataSampleAttachmentContent**](doc//DataSampleApi.md#getdatasampleattachmentcontent) | Get attachment content of a DataSample
*DataSampleApi* | [**getDataSampleAttachmentDocument**](doc//DataSampleApi.md#getdatasampleattachmentdocument) | Get document metadata of a DataSample attachment
*DataSampleApi* | [**matchDataSample**](doc//DataSampleApi.md#matchdatasample) | Find data samples ids using the provided Filter.
*DataSampleApi* | [**setDataSampleAttachment**](doc//DataSampleApi.md#setdatasampleattachment) | Add or update the attachment of a DataSample
*HealthcareElementApi* | [**createOrModifyHealthcareElement**](doc//HealthcareElementApi.md#createormodifyhealthcareelement) | Create a Healthcare Element
*HealthcareElementApi* | [**createOrModifyHealthcareElements**](doc//HealthcareElementApi.md#createormodifyhealthcareelements) | Create a Healthcare Element
*HealthcareElementApi* | [**deleteHealthcareElement**](doc//HealthcareElementApi.md#deletehealthcareelement) | Delete a Healthcare Element
*HealthcareElementApi* | [**filterHealthcareElement**](doc//HealthcareElementApi.md#filterhealthcareelement) | Find Healthcare Elements using a filter
*HealthcareElementApi* | [**getHealthcareElement**](doc//HealthcareElementApi.md#gethealthcareelement) | Get a Healthcare Element
*HealthcareElementApi* | [**matchHealthcareElement**](doc//HealthcareElementApi.md#matchhealthcareelement) | Find Healthcare Elements using a filter
*HealthcareProfessionalApi* | [**createOrModifyHealthcareProfessional**](doc//HealthcareProfessionalApi.md#createormodifyhealthcareprofessional) | Create a newhealthcare professional or modify an existing one.
*HealthcareProfessionalApi* | [**deleteHealthcareProfessional**](doc//HealthcareProfessionalApi.md#deletehealthcareprofessional) | Delete an existing healthcare professional.
*HealthcareProfessionalApi* | [**filterHealthcareProfessionalBy**](doc//HealthcareProfessionalApi.md#filterhealthcareprofessionalby) | Load healthcare professionals from the database by filtering them using the provided Filter.
*HealthcareProfessionalApi* | [**getHealthcareProfessional**](doc//HealthcareProfessionalApi.md#gethealthcareprofessional) | Get a Healthcare professional by id.
*HealthcareProfessionalApi* | [**matchHealthcareProfessionalBy**](doc//HealthcareProfessionalApi.md#matchhealthcareprofessionalby) | Loadhealthcare professional ids from the database by filtering them using the provided Filter.
*MedicalDeviceApi* | [**createOrModifyMedicalDevice**](doc//MedicalDeviceApi.md#createormodifymedicaldevice) | Create or update a [MedicalDevice]
*MedicalDeviceApi* | [**createOrModifyMedicalDevices**](doc//MedicalDeviceApi.md#createormodifymedicaldevices) | Create or update a batch of [MedicalDevice]
*MedicalDeviceApi* | [**deleteMedicalDevice**](doc//MedicalDeviceApi.md#deletemedicaldevice) | Delete a [MedicalDevice]
*MedicalDeviceApi* | [**deleteMedicalDevices**](doc//MedicalDeviceApi.md#deletemedicaldevices) | Delete a batch of [MedicalDevice]
*MedicalDeviceApi* | [**filterMedicalDevices**](doc//MedicalDeviceApi.md#filtermedicaldevices) | Load devices from the database by filtering them using the provided [filter].
*MedicalDeviceApi* | [**getMedicalDevice**](doc//MedicalDeviceApi.md#getmedicaldevice) | Get a Medical Device
*MedicalDeviceApi* | [**matchMedicalDevices**](doc//MedicalDeviceApi.md#matchmedicaldevices) | Load medical device ids from the database by filtering them using the provided Filter.
*PatientApi* | [**createOrModifyPatient**](doc//PatientApi.md#createormodifypatient) | Create or update a [Patient]
*PatientApi* | [**deletePatient**](doc//PatientApi.md#deletepatient) | Delete a [Patient]
*PatientApi* | [**filterPatients**](doc//PatientApi.md#filterpatients) | Load patients from the database by filtering them using the provided [filter].
*PatientApi* | [**getPatient**](doc//PatientApi.md#getpatient) | Get a [Patient]
*PatientApi* | [**matchPatients**](doc//PatientApi.md#matchpatients) | Load patient ids from the database by filtering them using the provided [filter].
*UserApi* | [**checkTokenValidity**](doc//UserApi.md#checktokenvalidity) | Check token validity for a user.
*UserApi* | [**createOrModifyUser**](doc//UserApi.md#createormodifyuser) | Create a new user or modify an existing one.
*UserApi* | [**createToken**](doc//UserApi.md#createtoken) | Create a token for a user.
*UserApi* | [**deleteUser**](doc//UserApi.md#deleteuser) | Delete an existing user.
*UserApi* | [**filterUsers**](doc//UserApi.md#filterusers) | Load users from the database by filtering them using the provided Filter.
*UserApi* | [**getLoggedUser**](doc//UserApi.md#getloggeduser) | Get the details of the logged User.
*UserApi* | [**getUser**](doc//UserApi.md#getuser) | Get a User by id.
*UserApi* | [**matchUsers**](doc//UserApi.md#matchusers) | Load user ids from the database by filtering them using the provided Filter.


## Documentation For Models

 - [Address](doc//Address.md)
 - [AuthenticationToken](doc//AuthenticationToken.md)
 - [Coding](doc//Coding.md)
 - [CodingReference](doc//CodingReference.md)
 - [Content](doc//Content.md)
 - [DataSample](doc//DataSample.md)
 - [Delegation](doc//Delegation.md)
 - [Document](doc//Document.md)
 - [Filter](doc//Filter.md)
 - [HealthcareElement](doc//HealthcareElement.md)
 - [HealthcareProfessional](doc//HealthcareProfessional.md)
 - [Identifier](doc//Identifier.md)
 - [InlineResponse403](doc//InlineResponse403.md)
 - [Measure](doc//Measure.md)
 - [MedicalDevice](doc//MedicalDevice.md)
 - [PaginatedDocumentKeyAndIdPairObject](doc//PaginatedDocumentKeyAndIdPairObject.md)
 - [PaginatedListCoding](doc//PaginatedListCoding.md)
 - [PaginatedListDataSample](doc//PaginatedListDataSample.md)
 - [PaginatedListHealthcareElement](doc//PaginatedListHealthcareElement.md)
 - [PaginatedListHealthcareProfessional](doc//PaginatedListHealthcareProfessional.md)
 - [PaginatedListMedicalDevice](doc//PaginatedListMedicalDevice.md)
 - [PaginatedListPatient](doc//PaginatedListPatient.md)
 - [PaginatedListUser](doc//PaginatedListUser.md)
 - [Partnership](doc//Partnership.md)
 - [Patient](doc//Patient.md)
 - [PatientHealthCareParty](doc//PatientHealthCareParty.md)
 - [PersonName](doc//PersonName.md)
 - [Property](doc//Property.md)
 - [PropertyType](doc//PropertyType.md)
 - [SystemMetaDataEncrypted](doc//SystemMetaDataEncrypted.md)
 - [SystemMetaDataOwner](doc//SystemMetaDataOwner.md)
 - [SystemMetaDataOwnerEncrypted](doc//SystemMetaDataOwnerEncrypted.md)
 - [Telecom](doc//Telecom.md)
 - [TimeSeries](doc//TimeSeries.md)
 - [TypedValueObject](doc//TypedValueObject.md)
 - [User](doc//User.md)


## Documentation For Authorization

 All endpoints do not require authorization.


## Author


CHANGELOG.md
# 1.5.4

- Interface for potentially encrypted patient now includes setters
- More api methods can use potentially encrypted patient

# 1.5.3

- Fix a bug that caused corruption of encrypted data on modification of unencrypted patient data.

# 1.5.2

- Fix a bug that caused corruption of the Patient entity data the first time the new patient user logins through the
  authentication API.
- Includes new methods that allow new Patient users which didn't yet receive access to their existing data to still be
  able to create new data samples and healthcare elements

# 1.5.1

- [MDT-175](https://icure.atlassian.net/browse/MDT-175) BUG : Resolve 400 Invalid Patient bug during patient
  authentication
- BUG : Resolve 500 Base64 error due to defected cache when completing patient authentication
- BUG : Resolve race condition when calling updateUser service

# 1.5.0

- Add Check to bypass token in case a user need to use authentication instead (Google Store submission)
- Enable authentication by SMS

# 1.4.9

- Fixed an error when completing authentication

# 1.4.8

- Bumped icure-dart-sdk
- Changed TimeSeries to list of num

# 1.4.7

- Fixed buggy services
- Fixed issue blocking registration

# 1.4.6

- Updated dart sdk

# 1.4.5

- FIX an issue on registration

# 1.4.4

- FEATURE Send recaptchaKey as sha256 during authentication process

# 1.4.3

- FIX an issue that inverted the logic of delegation sharing when the delegation is already existing

# 1.4.2

- FIX giveAccessTo wouldn't check if the delegateTo already has access
- FIX User ID is used in giveAccessTo in PatientApi
- FIX Cache could sometimes return wrong values for delegateHcPartyKeys

# 1.4.1

- Fix an issue that block the decryption of data samples

# 1.4.0

- Add the ability to give access to other data owner to healthcare element, data sample and patient
- Fix issues on iterable comparison
- Fix an issue on UserDto to User mapping

# 1.3.5

- BUG Fix missing identifier value mapping

# 1.3.4

- Add parameter to provide keys and tokens using a Keystore

# 1.3.3

- Add AnonymousMedtechApi for initialisation of users

# 1.3.2

- Renamed registrationApi -> authenticationApi

# 1.3.1

- Implement authentication login with user still having access to its key

# 1.3.0

- Added login process methods

# 1.2.17

- FEATURE Support AesExchangeKeys, transferKeys and lostHcPartyKeys in data owners
- Updating icure_dart_sdk dependency to 1.0.34

# 1.2.16

- FIX Add patientId in createOrModifyHealthElement API service
- Updating icure_dart_sdk dependency to 1.0.33

# 1.2.15

- Fixed creation of DataSamples that contains "content" value
- Updated icure-dart-sdk to 1.0.32

## 1.2.14

- FIX Could not manage data samples when user is a patient
- FIX Could not filter patients and healthcare elements for a patient data owner
- FIX HealthcareElementByHcPartyIdentifiersFilter mapping

## 1.2.13

- Updated icure dart sdk

## 1.2.12

- Updated icure dart sdk

## 1.2.11

- Fix createOrModifyHealthcareElements method
- Add missing methods in healthcare professional API and delete unrelated documentation
- Simplify MedTechApiBuilder

## 1.2.10

- Fix registration process
-

## 1.2.9

- Added the possibility to provide a validationCode for the signUp

## 1.2.8

- Fixed issues in Filters DSL

## 1.2.7

- Exported binary_utils
- Changed addKeyPair signature

## 1.2.6

- Registration API Use Client Process ID

## 1.2.5

- Fix jsonEncode error when calling registration api

## 1.2.4

- Changed forHcp to forDataOwner in filter_dsl

## 1.2.3

- Added simplified calls for registration

## 1.2.2

- Exported crypto from icure-dart-sdk

## 1.2.1

- Update dependencies

## 1.2.0

- Added getUserByEmail

## 1.1.5

- Update dependencies & add signup test

## 1.1.4

- Update dependencies

## 1.1.3

- Update dependencies

## 1.1.2

- Fix forPatient DataSample filter builder

## 1.1.1

- Update dependencies

## 1.1.0

- Added data samples API

## 1.0.7

- Better documentation, better top level api

## 1.0.6

- Performance improvement

## 1.0.5

- Fix transitive dependency problem

## 1.0.4

- Improve library layout

## 1.0.3

- Fix filter serialization, mapping and decryption of entities without encryption keys

## 1.0.2

- Fix filter mapping
- Add tests for healthcare professionals API

## 1.0.1

- Add healthcare elements API
- Add healthcare professionals API
- Improve test coverage
- Add constructors to filters

## 1.0.0

- Initial version


lib/api_client.dart
//
// AUTO-GENERATED FILE, DO NOT MODIFY!
//
// @dart=2.12

// ignore_for_file: unused_element, unused_import
// ignore_for_file: always_put_required_named_parameters_first
// ignore_for_file: constant_identifier_names
// ignore_for_file: lines_longer_than_80_chars

part of icure_medical_device_dart_sdk.api;

class ApiClient {
  ApiClient({this.basePath = 'http://127.0.0.1:8912'}) {
  }

  final String basePath;

  var _client = Client();

  /// Returns the current HTTP [Client] instance to use in this class.
  ///
  /// The return value is guaranteed to never be null.
  Client get client => _client;

  /// Requests to use a new HTTP [Client] in this class.
  set client(Client newClient) {
    _client = newClient;
  }

  final _defaultHeaderMap = <String, String>{};
  final _authentications = <String, Authentication>{};

  void addDefaultHeader(String key, String value) {
     _defaultHeaderMap[key] = value;
  }

  Map<String,String> get defaultHeaderMap => _defaultHeaderMap;

  /// Returns an unmodifiable [Map] of the authentications, since none should be added
  /// or deleted.
  Map<String, Authentication> get authentications => Map.unmodifiable(_authentications);

  T? getAuthentication<T extends Authentication>(String name) {
    final authentication = _authentications[name];
    return authentication is T ? authentication : null;
  }

  // We don't use a Map<String, String> for queryParams.
  // If collectionFormat is 'multi', a key might appear multiple times.
  Future<Response> invokeAPI(
    String path,
    String method,
    List<QueryParam> queryParams,
    Object? body,
    Map<String, String> headerParams,
    Map<String, String> formParams,
    String? contentType,
    List<String> authNames,
  ) async {
    _updateParamsForAuth(authNames, queryParams, headerParams);

    headerParams.addAll(_defaultHeaderMap);
    if (contentType != null) {
      headerParams['Content-Type'] = contentType;
    }

    final urlEncodedQueryParams = queryParams.map((param) => '$param');
    final queryString = urlEncodedQueryParams.isNotEmpty ? '?${urlEncodedQueryParams.join('&')}' : '';
    final uri = Uri.parse('$basePath$path$queryString');

    try {
      // Special case for uploading a single file which isn't a 'multipart/form-data'.
      if (
        body is MultipartFile && (contentType == null ||
        !contentType.toLowerCase().startsWith('multipart/form-data'))
      ) {
        final request = StreamedRequest(method, uri);
        request.headers.addAll(headerParams);
        request.contentLength = body.length;
        body.finalize().listen(
          request.sink.add,
          onDone: request.sink.close,
          // ignore: avoid_types_on_closure_parameters
          onError: (Object error, StackTrace trace) => request.sink.close(),
          cancelOnError: true,
        );
        final response = await _client.send(request);
        return Response.fromStream(response);
      }

      if (body is MultipartRequest) {
        final request = MultipartRequest(method, uri);
        request.fields.addAll(body.fields);
        request.files.addAll(body.files);
        request.headers.addAll(body.headers);
        request.headers.addAll(headerParams);
        final response = await _client.send(request);
        return Response.fromStream(response);
      }

      final msgBody = contentType == 'application/x-www-form-urlencoded'
        ? formParams
        : await serializeAsync(body);
      final nullableHeaderParams = headerParams.isEmpty ? null : headerParams;

      switch(method) {
        case 'POST': return await _client.post(uri, headers: nullableHeaderParams, body: msgBody,);
        case 'PUT': return await _client.put(uri, headers: nullableHeaderParams, body: msgBody,);
        case 'DELETE': return await _client.delete(uri, headers: nullableHeaderParams, body: msgBody,);
        case 'PATCH': return await _client.patch(uri, headers: nullableHeaderParams, body: msgBody,);
        case 'HEAD': return await _client.head(uri, headers: nullableHeaderParams,);
        case 'GET': return await _client.get(uri, headers: nullableHeaderParams,);
      }
    } on SocketException catch (error, trace) {
      throw ApiException.withInner(
        HttpStatus.badRequest,
        'Socket operation failed: $method $path',
        error,
        trace,
      );
    } on TlsException catch (error, trace) {
      throw ApiException.withInner(
        HttpStatus.badRequest,
        'TLS/SSL communication failed: $method $path',
        error,
        trace,
      );
    } on IOException catch (error, trace) {
      throw ApiException.withInner(
        HttpStatus.badRequest,
        'I/O operation failed: $method $path',
        error,
        trace,
      );
    } on ClientException catch (error, trace) {
      throw ApiException.withInner(
        HttpStatus.badRequest,
        'HTTP connection failed: $method $path',
        error,
        trace,
      );
    } on Exception catch (error, trace) {
      throw ApiException.withInner(
        HttpStatus.badRequest,
        'Exception occurred: $method $path',
        error,
        trace,
      );
    }

    throw ApiException(
      HttpStatus.badRequest,
      'Invalid HTTP operation: $method $path',
    );
  }

  Future<dynamic> deserializeAsync(String json, String targetType, {bool growable = false,}) async =>
    // ignore: deprecated_member_use_from_same_package
    deserialize(json, targetType, growable: growable);

  @Deprecated('Scheduled for removal in OpenAPI Generator 6.x. Use deserializeAsync() instead.')
  dynamic deserialize(String json, String targetType, {bool growable = false,}) {
    // Remove all spaces. Necessary for regular expressions as well.
    targetType = targetType.replaceAll(' ', ''); // ignore: parameter_assignments

    // If the expected target type is String, nothing to do...
    return targetType == 'String'
      ? json
      : _deserialize(jsonDecode(json), targetType, growable: growable);
  }

  // ignore: deprecated_member_use_from_same_package
  Future<String> serializeAsync(Object? value) async => serialize(value);

  @Deprecated('Scheduled for removal in OpenAPI Generator 6.x. Use serializeAsync() instead.')
  String serialize(Object? value) => value == null ? '' : json.encode(value);

  /// Update query and header parameters based on authentication settings.
  /// @param authNames The authentications to apply
  void _updateParamsForAuth(
    List<String> authNames,
    List<QueryParam> queryParams,
    Map<String, String> headerParams,
  ) {
    for(final authName in authNames) {
      final auth = _authentications[authName];
      if (auth == null) {
        throw ArgumentError('Authentication undefined: $authName');
      }
      auth.applyToParams(queryParams, headerParams);
    }
  }

  static dynamic _deserialize(dynamic value, String targetType, {bool growable = false}) {
    try {
      switch (targetType) {
        case 'String':
          return value is String ? value : value.toString();
        case 'int':
          return value is int ? value : int.parse('$value');
        case 'double':
          return value is double ? value : double.parse('$value');
        case 'bool':
          if (value is bool) {
            return value;
          }
          final valueString = '$value'.toLowerCase();
          return valueString == 'true' || valueString == '1';
        case 'Address':
          return Address.fromJson(value);
        case 'AuthenticationToken':
          return AuthenticationToken.fromJson(value);
        case 'Coding':
          return Coding.fromJson(value);
        case 'CodingReference':
          return CodingReference.fromJson(value);
        case 'Content':
          return Content.fromJson(value);
        case 'DataSample':
          return DataSample.fromJson(value);
        case 'Delegation':
          return Delegation.fromJson(value);
        case 'Document':
          return Document.fromJson(value);
        case 'HealthcareElement':
          return HealthcareElement.fromJson(value);
        case 'HealthcareProfessional':
          return HealthcareProfessional.fromJson(value);
        case 'Identifier':
          return Identifier.fromJson(value);
        case 'InlineResponse403':
          return InlineResponse403.fromJson(value);
        case 'Measure':
          return Measure.fromJson(value);
        case 'MedicalDevice':
          return MedicalDevice.fromJson(value);
        case 'PaginatedDocumentKeyAndIdPairObject':
          return PaginatedDocumentKeyAndIdPairObject.fromJson(value);
        case 'PaginatedListCoding':
          return PaginatedListCoding.fromJson(value);
        case 'PaginatedListDataSample':
          return PaginatedListDataSample.fromJson(value);
        case 'PaginatedListHealthcareElement':
          return PaginatedListHealthcareElement.fromJson(value);
        case 'PaginatedListHealthcareProfessional':
          return PaginatedListHealthcareProfessional.fromJson(value);
        case 'PaginatedListMedicalDevice':
          return PaginatedListMedicalDevice.fromJson(value);
        case 'PaginatedListPatient':
          return PaginatedListPatient.fromJson(value);
        case 'PaginatedListUser':
          return PaginatedListUser.fromJson(value);
        case 'Partnership':
          return Partnership.fromJson(value);
        case 'Patient':
          return Patient.fromJson(value);
        case 'PatientHealthCareParty':
          return PatientHealthCareParty.fromJson(value);
        case 'PersonName':
          return PersonName.fromJson(value);
        case 'Property':
          return Property.fromJson(value);
        case 'PropertyType':
          return PropertyType.fromJson(value);
        case 'SystemMetaDataEncrypted':
          return SystemMetaDataEncrypted.fromJson(value);
        case 'SystemMetaDataOwner':
          return SystemMetaDataOwner.fromJson(value);
        case 'SystemMetaDataOwnerEncrypted':
          return SystemMetaDataOwnerEncrypted.fromJson(value);
        case 'Telecom':
          return Telecom.fromJson(value);
        case 'TimeSeries':
          return TimeSeries.fromJson(value);
        case 'TypedValueObject':
          return TypedValueObject.fromJson(value);
        case 'User':
          return User.fromJson(value);
        default:
          dynamic match;
          if (value is List && (match = _regList.firstMatch(targetType)?.group(1)) != null) {
            return value
              .map<dynamic>((dynamic v) => _deserialize(v, match, growable: growable,))
              .toList(growable: growable);
          }
          if (value is Set && (match = _regSet.firstMatch(targetType)?.group(1)) != null) {
            return value
              .map<dynamic>((dynamic v) => _deserialize(v, match, growable: growable,))
              .toSet();
          }
          if (value is Map && (match = _regMap.firstMatch(targetType)?.group(1)) != null) {
            return Map<String, dynamic>.fromIterables(
              value.keys.cast<String>(),
              value.values.map<dynamic>((dynamic v) => _deserialize(v, match, growable: growable,)),
            );
          }
      }
    } on Exception catch (error, trace) {
      throw ApiException.withInner(HttpStatus.internalServerError, 'Exception during deserialization.', error, trace,);
    }
    throw ApiException(HttpStatus.internalServerError, 'Could not find a suitable class for deserialization',);
  }
}

/// Primarily intended for use in an isolate.
class DeserializationMessage {
  const DeserializationMessage({
    required this.json,
    required this.targetType,
    this.growable = false,
  });

  /// The JSON value to deserialize.
  final String json;

  /// Target type to deserialize to.
  final String targetType;

  /// Whether to make deserialized lists or maps growable.
  final bool growable;
}

/// Primarily intended for use in an isolate.
Future<dynamic> deserializeAsync(DeserializationMessage message) async {
  // Remove all spaces. Necessary for regular expressions as well.
  final targetType = message.targetType.replaceAll(' ', '');

  // If the expected target type is String, nothing to do...
  return targetType == 'String'
    ? message.json
    : ApiClient._deserialize(
        jsonDecode(message.json),
        targetType,
        growable: message.growable,
      );
}

/// Primarily intended for use in an isolate.
Future<String> serializeAsync(Object? value) async => value == null ? '' : json.encode(value);
